llvm/lib/CodeGen/PseudoSourceValue.cpp

```diff
@@ -122,7 +122,12 @@ const PseudoSourceValue *PseudoSourceValueManager::getJumpTable() {
 
 const PseudoSourceValue *
 PseudoSourceValueManager::getFixedStack(int FI) {
-  std::unique_ptr<FixedStackPseudoSourceValue> &V = FSValues[FI];
+  // Frame index is often continuously positive, but can be negative. Use
+  // zig-zag encoding for dense index into FSValues vector.
+  unsigned Idx = FI >= 0 ? FI << 1 : (-FI - 1) << 1 | 1;
```

Comment:
Use `(2*FI) ^ (FI >> (sizeof(FI)*8-1));`? The generated code is more efficient

Reply:
Good point, changed.